### PR TITLE
Fix api#201 - Bug 1

### PR DESCRIPTION
**Pull request by Dunky13**
Fix redkubes/otomi-api#201 bug number 1, as stated in https://github.com/redkubes/otomi-api/issues/201#issuecomment-901864370

Where encrypting/decrypting would take up too much resources when do all files in parallel.
Now it is chunked up into smaller parallel executions.

### CODE REVIEW

File: src/common/crypt.ts [hunk 1]
```diff
@@ -51,21 +66,20 @@ const runOnSecretFiles = async (crypt: CR, filesArgs: string[] = []): Promise<Pr
     files = await getAllSecretFiles()
   }
   preCrypt()
+  const chunkSize = 5
+  const filesChunked = chunkArray(files, chunkSize)
+
   const eventEmitterDefaultListeners = EventEmitter.defaultMaxListeners
-  EventEmitter.defaultMaxListeners = files.length + 5
+  if (chunkSize + 2 > EventEmitter.defaultMaxListeners) EventEmitter.defaultMaxListeners = chunkSize + 2
```
Comment (j-zimnowoda) [inline src/common/crypt.ts:73]: it is not clear why do you update `EventEmitter.defaultMaxListeners`

File: src/common/envalid.ts [hunk 2]
```diff
@@ -1,6 +1,21 @@
-import { config } from 'dotenv'
+import { parse } from 'dotenv'
 import { bool, cleanEnv, json, str } from 'envalid'
-import { existsSync } from 'fs'
+import { existsSync, readFileSync } from 'fs'
+
+export const dotEnvParse = (path: string): void => {
```
Comment (j-zimnowoda) [inline src/common/envalid.ts:5]: I thought that dotEnvParse should be part of different PR right ?

File: src/common/crypt.ts [hunk 3]
```diff
@@ -51,21 +66,31 @@ const runOnSecretFiles = async (crypt: CR, filesArgs: string[] = []): Promise<Pr
     files = await getAllSecretFiles()
   }
   preCrypt()
+  const CHUNK_SIZE = 5
```
[the remark this answers is not in the document]
Comment (Dunky13) [reply src/common/crypt.ts]: Already done ^_^

File: src/common/crypt.ts (hunk 1 above)
Comment (svatwork) [inline src/common/crypt.ts:74]: Why not say chunkSize = 7 instead of adding 2?

File: src/common/envalid.ts (hunk 2 above)
Comment (Dunky13) [reply src/common/envalid.ts:5]: Whoops, branched from wrong branch, thanks for helping me fix it :)

**Review by svatwork — CHANGES_REQUESTED** [review]
I see the original problem was not enough resources. Of course we could disable parallel execution, divide up the workload in chunks but why not increase available resources? What order of magnitude are we talking about here of secrets that we need to decrypt that it would not have enough memory? 

Seems like the problem will reappear if the node doesn't have enough resources, e.g. in a KinD.

File: src/common/crypt.ts (hunk 1 above)
Comment (Dunky13) [reply src/common/crypt.ts:74]: Adding 2 to make sure that there is room for other parallel execution to be run

**Review by j-zimnowoda — APPROVED** [review]

**Review by mojtabaimani — APPROVED** [review]

File: src/common/crypt.ts (hunk 1 above)
Comment (Dunky13) [reply src/common/crypt.ts:74]: I could implement a parallel queue, as suggested in this article: https://techsparx.com/nodejs/async/queue-processing.html

Which should be able to queue, but afaik (and I've done a masters in CS Parallel Programming - so I should know (I hope)) the developer always needs to make a decision / calculate the resources available.

So here I've decided arbitrarily to chunk the work per-5 basis (and add 2 parallel threads for overhead)
If I use a queue, the same thing basically happens, maybe sightly smarted that full on chunks. But in that article you also read that `concurren(t/cy)` needs to be set (they set it to 5 as well)

I could find how many CPU cores are available and use that - 1 or -2 (need some cores idle to be able to process the queue, results and other system resources) But if decryptions is indeed memory bound (I found that decrypting 5 files takes 40MB in execution - so I can't imagine it being memory bound) then I need to figure out how much decryption takes in memory, divide available memory, calculate and make an estimation. But then again, that also wastes computational cycles, where I (as a developer) need to make an assumption how much memory a single decryption takes.

As you can read in this write up, in parallel programming there are a lot of nuances to think of if you want to optimize of most efficient. But sometime parallel programming can be made easier by making hardcoded arbitrary assumptions (like 5 decryptions in parallel) because it is "good enough"

Now I'm happy to optimize this some more, and doing some calcuations, but I think that 5 would be a nice limit.

Another thing you mentioned, disabling parallel execution is definitely an option, but rewriting the CLI from bash to TS, was that it is easier tot parallelize execution. One of the use cases for parallel execution was this exact function en/decryptions of secrets file. So disabling parallelization means we lose the performance gain here. Which would be a shame tbh, especially since this seems to work. If you find that during KinD there are not enough resources, then I suggest we look into optimizing the chunkSize instead of disabling parallelization completely. Since a chunkSize of 1 would be sequential anyway.

**Review by svatwork — APPROVED** [review]
